### PR TITLE
Merge CRC32.h into Hash.h and remove MD5.h

### PR DESCRIPTION
**Pull request by Pokechu22**
The main point of this is to have one obvious location where all hash-related code is.  I've also used the wrapper around zlib's `crc32` in other places (it was first added for the GameCube IPL rom in 3d662e746bef6651dcdd0bc985c9e6d6518f006d).

### CODE REVIEW

File: Source/Core/Common/Hash.h [hunk 1]
```diff
@@ -4,14 +4,28 @@
 #pragma once
 
 #include <cstddef>
+#include <functional>
+#include <string_view>
 
 #include "Common/CommonTypes.h"
 
+namespace DiscIO
```
Comment (lioncash) [inline Source/Core/Common/Hash.h:12]: Common shouldn't be dependent on DiscIO

**Review by lioncash — CHANGES_REQUESTED** [review]

File: Source/Core/Common/Hash.h (hunk 1 above)
Comment (Pokechu22) [reply Source/Core/Common/Hash.h:12]: MD5.cpp already was in common and depended on DiscIO.

Comment (JosJuice) [reply Source/Core/Common/Hash.h:12]: The only function in MD5.cpp seems to have NetPlayClient.cpp as its only caller. I would suggest moving it either to there or to DiscIO.

File: Source/Core/Core/NetPlayClient.cpp [hunk 2]
```diff
@@ -2437,6 +2436,43 @@ bool NetPlayClient::DoAllPlayersHaveGame()
   });
 }
 
+static std::string MD5Sum(const std::string& file_path, std::function<bool(int)> report_progress)
```
Comment (JosJuice) [inline Source/Core/Core/NetPlayClient.cpp:2439]: `#include <functional>`

File: Source/Core/Core/NetPlayClient.cpp [hunk 3]
```diff
@@ -2437,6 +2437,43 @@ bool NetPlayClient::DoAllPlayersHaveGame()
...
+  }
+
+  std::array<u8, 16> output;
+  mbedtls_md5_finish_ret(&ctx, output.data());
+
+  // Convert to hex
+  for (u8 n : output)
+    output_string += fmt::format("{:02x}", n);
+
+  return output_string;
```
Comment (iwubcode) [inline Source/Core/Core/NetPlayClient.cpp:2474]: `return fmt::format("{:02x}", fmt::join(output, ""));` ?

Comment (Pokechu22) [reply Source/Core/Core/NetPlayClient.cpp]: Changed - I don't think this is necessarily more performant, but it's also not going to be significantly worse.  (It's formatting an array of 16 characters, and this happens once per netplay session.)

File: Source/Core/Core/NetPlayClient.cpp [hunk 4]
```diff
@@ -2462,6 +2462,43 @@ bool NetPlayClient::DoAllPlayersHaveGame()
...
 }
 
+static std::string MD5Sum(const std::string& file_path, std::function<bool(int)> report_progress)
+{
+  std::string output_string;
+  std::vector<u8> data(8 * 1024 * 1024);
+  u64 read_offset = 0;
+  mbedtls_md5_context ctx;
+
+  std::unique_ptr<DiscIO::BlobReader> file(DiscIO::CreateBlobReader(file_path));
```
Comment (leoetlino) [inline Source/Core/Core/NetPlayClient.cpp:2471]: Missing include for DiscIO::CreateBlobReader (DiscIO/Blob.h).

**Review by leoetlino — CHANGES_REQUESTED** [review]

**Review by leoetlino — APPROVED** [review]

**Pull request merged**